src/tracing.ts

```diff
@@ -0,0 +1,173 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as common from '@google-cloud/common';
+import * as path from 'path';
+import * as semver from 'semver';
+
+import {cls, TraceCLSConfig, TraceCLSMechanism} from './cls';
+import {CLSMechanism, Config, defaultConfig} from './config';
+import {Constants} from './constants';
+import * as PluginTypes from './plugin-types';
+import {TraceAgent} from './trace-api';
+import {pluginLoader, PluginLoaderConfig} from './trace-plugin-loader';
+import {traceWriter, TraceWriterConfig} from './trace-writer';
+import {Component, FORCE_NEW, Forceable, packageNameFromPath, Singleton} from './util';
+
+export interface TopLevelConfig {
+  enabled: boolean;
+  logLevel: number;
+  clsMechanism: CLSMechanism;
+}
+
+// PluginLoaderConfig extends TraceAgentConfig
+export type NormalizedConfig =
+    (TraceWriterConfig&PluginLoaderConfig&TopLevelConfig)|{enabled: false};
+
+/**
+ * A class that represents automatic tracing.
+ */
+export class Tracing implements Component {
+  /** An object representing the custom span API. */
+  private readonly traceAgent: TraceAgent;
+  /** A logger. */
+  private readonly logger: common.Logger;
+  /** The configuration object for this instance. */
+  private readonly config: Forceable<NormalizedConfig>;
+
+  /**
+   * Constructs a new Tracing instance.
+   * @param config The configuration for this instance.
+   */
+  constructor(config: NormalizedConfig, traceAgent: TraceAgent) {
```

Comment:
nit: make these `private readonly` here, remove assignments below and declarations above.

Reply:
The type of `config` is a little different (`this.config: Forceable<T>`; `config: T`). I moved `traceAgent`.